pennylane/gradients/general_shift_rules.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a function for generating generalized parameter shift rules."""
+import numpy as np
+
+
+def get_shift_rule(frequencies, diff_shifts=None):
+    r"""Computes the parameter shift rule for a unitary based on its generator's eigenvalue frequency
+     spectrum.
+
+    To compute gradients of circuit parameters in variational quantum algorithms, expressions for
+    cost function first derivatives with respect to the variational parameters can be cast into
+    linear combinations of expectation values at shifted parameter values. These "gradient recipes"
+    can be obtained from the unitary generator's eigenvalue frequency spectrum. Details can be
+    found in https://arxiv.org/abs/2107.12390.
+
+    Args:
+        frequencies (list): the list of eigenvalue frequencies. Eigenvalue frequencies are defined
+            as the unique positive differences obtained from a set of eigenvalues.
+        diff_shifts (list): the list of shift values. If unspecified, equidistant shifts are

```

Comment:
I opted for `diff_shifts` to emphasize it is the list of shifts used in the _differences_ in the gradient reconstruction, i.e.
`diff_shifts = [s_1, s_2] gives a gradient rule of the form

`c_1 * f(x + s_1) - c_1 * f(x - s_1) + c_2 * f(x + s_2) - c_2 *f(x - s_2)`,  rather than just

`c_1 * f(x + s_1) + c_2 * f(x + s_2)`.

However for brevity it may make sense to rename `diff_shifts` to `shifts` and make this functionality apparent in the docstring. I've taken this approach in the latest commit.